sdk/eventhub/event-hubs/src/eventHubReceiver.ts

```diff
@@ -676,153 +523,222 @@ export class EventHubReceiver extends LinkEntity {
+    cleanupBeforeAbort?: () => void;
+    abortErrorMsg?: string;
+  }
+): Promise<void> {
+  let token: ReturnType<typeof setTimeout>;
+  return createAbortablePromise<void>((resolve) => {
+    token = setTimeout(resolve, waitTimeInMs);
+  }, options).finally(() => clearTimeout(token));
+}
 
-    const config: RetryConfig<ReceivedEventData[]> = Object.defineProperties(
-      {
-        operation: retrieveEvents,
-        operationType: RetryOperationType.receiveMessage,
-        abortSignal: abortSignal,
-        retryOptions: retryOptions,
-      },
-      {
-        connectionId: {
-          enumerable: true,
-          get: () => {
-            return this._context.connectionId;
-          },
-        },
-        connectionHost: {
-          enumerable: true,
-          get: () => {
-            return this._context.config.host;
-          },
-        },
+export function checkOnInterval(
+  waitTimeInMs: number,
+  check: () => boolean,
+  options?: {
+    abortSignal?: AbortSignalLike;
+    cleanupBeforeAbort?: () => void;
+    abortErrorMsg?: string;
+  }
+): Promise<void> {
+  let token: ReturnType<typeof setInterval>;
+  return createAbortablePromise<void>((resolve) => {
+    token = setInterval(() => {
+      if (check()) {
+        resolve();
       }
-    ) as RetryConfig<ReceivedEventData[]>;
-    return retry<ReceivedEventData[]>(config);
+    }, waitTimeInMs);
+  }, options).finally(() => clearInterval(token));
+}
+
+/**
+ * Returns a promise that will resolve when it is time to read from the queue
+ * @param maxEventCount - The maximum number of events to receive.
+ * @param maxWaitTimeInMs - The maximum time to wait in ms for the queue to contain any events.
+ * @param readIntervalWaitTimeInMs - The time interval to wait in ms before checking the queue.
+ * @param queue - The queue to read from.
+ * @param options - The options bag.
+ * @returns a promise that will resolve when it is time to read from the queue
+ */
+export function waitForEvents(
+  maxEventCount: number,
+  maxWaitTimeInMs: number,
+  readIntervalWaitTimeInMs: number,
+  queue: unknown[],
+  options?: {
+    abortSignal?: AbortSignalLike;
+    cleanupBeforeAbort?: () => void;
+    receivedAfterWait?: (count: number) => void;
+    receivedAlready?: () => void;
+    receivedNone?: () => void;
   }
+): Promise<void> {
+  const {
+    abortSignal: clientAbortSignal,
+    cleanupBeforeAbort,
+    receivedNone,
+    receivedAfterWait,
+    receivedAlready,
+  } = options ?? {};
+  const aborter = new AbortController();
+  const { signal: abortSignal } = new AbortController([
+    aborter.signal,
+    ...(clientAbortSignal ? [clientAbortSignal] : []),
+  ]);
+  const updatedOptions = {
+    abortSignal,
+    abortErrorMsg: StandardAbortMessage,
+    cleanupBeforeAbort: () => {
+      if (clientAbortSignal?.aborted) {
+        cleanupBeforeAbort?.();
+      }
+    },
+  };
+  return queue.length >= maxEventCount
+    ? Promise.resolve().then(receivedAlready)
```

Comment:
nitpick: `Promise.resolve(receivedAlready?.())` would work too?

Reply:
Yes, but I prefer the current line because it makes the intent clear it is a callback that gets called after the wait is over. While taking a look at this, I simplified the `receivedAfterWait` callback.